### PR TITLE
*: Remove retry limit and fix race conditions around ObjectFile

### PR DESCRIPTION
**Pull request by kakkoyun**
### Why?

There are two major fixes in this PR:

1) The Parca server sometimes restarts; if we don't ask the server constantly, we miss debuginfo files. Therefore, I'm removing the retry limit, and we will ask the server always if it has the necessary debuginfo.

2) Avoid any possible race conditions while reading the object files. 
Fixes https://github.com/parca-dev/parca-agent/issues/1614

### <samp>🤖 Generated by Copilot at a62d7f9</samp>

*  Add a new flag `--debuginfo-upload-max-parallel` to control the maximum number of parallel uploads of debuginfo files ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-57833d176ed681886b67a7fa2e4dc2f68d5e00cb1db269e7854fb3dafb45f54eL168-R168),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-57833d176ed681886b67a7fa2e4dc2f68d5e00cb1db269e7854fb3dafb45f54eL533-R534),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L79-R78),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L101-R100),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74L62-R71),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74R86-R310))
*  Refactor the `DebuginfoManager` interface and the `Manager` type to return the debuginfo object file instead of modifying the source object file, and to remove the retry logic and use a singleflight mechanism instead ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L18-R19),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L27),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L66),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L126-R123),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L142-R134),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L153-R147),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L168-R163),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L194-R196),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359R202),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L230-R218),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L281-R226),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L309-R256),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L329-L330),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359R277),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L350-R307),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L380-R339),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359R389),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359R395-R398),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L453-R414),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-4938d357026bd7ab52d9b4b374899dd4eced576aed8c33f9f871f6537bc03328L35),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-4938d357026bd7ab52d9b4b374899dd4eced576aed8c33f9f871f6537bc03328L70-L73),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7L36-R41),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7L198-R250),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7L231-R267))
*  Refactor the `ObjectFile` type to hide the underlying file descriptor and elf file, and to provide methods for concurrent access to the file with proper locking ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537R25),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L39-R56),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L69-R124),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537R132-R156),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L122-R171),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L153-R195),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L74-R83),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L87-R93),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L94-R102),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L106-R120),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L129-R136),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL214-R212),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL229-R222),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL257-R251),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL327-R326),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL353-R332),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-2fbc63cb8ebb2eae13c9bd2ff211fe4abf8a4a939fe97b804012f013d9a1a6eaL258-R259),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-2fbc63cb8ebb2eae13c9bd2ff211fe4abf8a4a939fe97b804012f013d9a1a6eaL336-R336),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-2fbc63cb8ebb2eae13c9bd2ff211fe4abf8a4a939fe97b804012f013d9a1a6eaL425-R425),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-38da4938004f35685ca8cb0368582c999d7db09d15a029522c945feff69e6ad5L119-R119),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-b3a7f294a0fa5a71ab06af71de938e5dfbc119109c6d13ce194e58efec847517R58-R62),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-b3a7f294a0fa5a71ab06af71de938e5dfbc119109c6d13ce194e58efec847517L71-R76))
*  Add a new type `SeekReaderAt` that combines the `io.ReaderAt` and `io.Seeker` interfaces, and modify the `Extract` function to take a `SeekReaderAt` instead of an `*os.File` as the source parameter ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-447c43aa1b1bd3f3896570c8d55826301501d359871dfda8fdd5c1e1ed193803R32-R36),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-447c43aa1b1bd3f3896570c8d55826301501d359871dfda8fdd5c1e1ed193803L74-R79),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-447c43aa1b1bd3f3896570c8d55826301501d359871dfda8fdd5c1e1ed193803L81-R86))
*  Modify the `readDebuglink` function to take an `*elf.File` instead of an `*os.File` as the parameter, and use the cached value if available ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L86-R86),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L140-R145),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L188-R190),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L205-R201),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-ce8f0dbf47ed5d0ee688ab01344a7de40c12a3690b890014a36fe073bd0fc865R18),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-ce8f0dbf47ed5d0ee688ab01344a7de40c12a3690b890014a36fe073bd0fc865L296-R300))
*  Modify the `Mapping` type to rename the `open` method to `init`, and to call the `computeKernelOffset` method from the `openObjFile` method ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL74-R75),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL134-R137),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL155-R154))
*  Add a new field `debuginfoSrcCache` to the `InfoManager` type, to cache the debuginfo object files by their build IDs ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7R120-R121),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7R136-R140),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-bc9c9211670f079eed672ee83ccef854706b1a639e4d9a0a28c92f5b4f5f0eb7L156-R173))
*  Modify the type of the `mtx` field of the `Manager` type in the `pkg/discovery` package to be a pointer to a `sync.RWMutex` instead of a value ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-13a577592bd80f3d398f9ea4848fd767e8a9677e1e64db96848d6700ac231055L89-R89),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-13a577592bd80f3d398f9ea4848fd767e8a9677e1e64db96848d6700ac231055R119))
*  Remove the `noopDebuginfoManager` type and its methods, as they are no longer needed ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-57833d176ed681886b67a7fa2e4dc2f68d5e00cb1db269e7854fb3dafb45f54eL916-R918))
*  Remove the import of the `github.com/cenkalti/backoff` package, as it is no longer used ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L8))
*  Add tests for the `Upload` method of the `Manager` type in the `pkg/debuginfo` package ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74L21-R28),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74L29-R37),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74R86-R310))
*  Modify the error message and the panic message in the `readDebuglink` function and the `pkg/objectfile/pool.go` file to include the type of the unexpected value in the cache ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L86-R86),[link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L129-R136))
*  Modify the condition to check for a successful status code in the `uploadViaHTTP` function to allow any 2xx status code instead of only 200 ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L453-R414))
*  Update the documentation of the `--debuginfo-upload-max-parallel` flag in the `README.md` file ([link](https://github.com/parca-dev/parca-agent/pull/1623/files?diff=unified&w=0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L112-R114))

### Test Plan

Relevant tests are added.

1. `make test`
1. Run `scripts/local-run.sh`
1. Run `make dev/up`

### <samp>🤖 Generated by Copilot at 07d7cda</samp>

> _Debuginfo manager_
> _Simplifies and improves code_
> _No more `backoff`_

### CODE REVIEW

**Comment by kakkoyun** [comment]
I will handle  https://github.com/parca-dev/parca-agent/issues/1617 in another PR

**Review by brancz — APPROVED** [review]

File: pkg/debuginfo/manager.go [hunk 1]
```diff
@@ -430,11 +386,16 @@ func (di *Manager) uploadFile(ctx context.Context, uploadInstructions *debuginfo
...
+	// NewGrpcUploadClient using bufio.NewReader to avoid closing the reader.
 	_, err := parcadebuginfo.NewGrpcUploadClient(debuginfoClient).Upload(ctx, uploadInstructions, r)
 	return err
 }
 
 func uploadViaSignedURL(ctx context.Context, url string, r io.Reader, size int64) error {
+	// Client is closing the reader if the reader is also closer.
+	// We need to wrap the reader to avoid this.
+	// We want to have total control over the reader.
+	r = bufio.NewReader(r)
```
Comment (kakkoyun) [inline pkg/debuginfo/manager.go:398]: This one was particularly annoying to track down.

**Review by brancz — APPROVED** [review]

**Comment by kakkoyun** [comment]
There's a race detected. I trying to understand if it's the actual code or the tests I have added.

**Comment by kakkoyun** [comment]
Also, this has been running since yesterday without a crash. That being said, I will still continue to improve with subsequent PRs.

File: README.md [hunk 2]
```diff
@@ -109,9 +109,9 @@ Flags:
       --debuginfo-strip            Only upload information needed for
                                    symbolization. If false the exact binary the
                                    agent sees will be uploaded unmodified.
-      --debuginfo-upload-retry-count=3
-                                   The number of times to retry uploading
-                                   debuginfo files.
+      --debuginfo-upload-max-parallel=25
```
Comment (javierhonduco) [inline README.md:112]: Glad this is configurable now! I wonder what the right default value is, perhaps 25 is very high? We've seen debuginfos that weight 100s of GB which could use too much bandwidth perhaps

File: pkg/debuginfo/manager.go (hunk 1 above)
Comment (javierhonduco) [reply pkg/debuginfo/manager.go:398]: Oh wow, that's a tricky one!

**Review by javierhonduco — APPROVED** [review]
Great work!! Love the tests and the manual tests you've ran to ensure this fixes the problem 😄

File: README.md (hunk 2 above)
Comment (kakkoyun) [reply README.md:112]: Happy to reduce the default value 💯

**Pull request merged**